src/hotspot/share/opto/subnode.hpp

```diff
@@ -297,6 +297,89 @@ class CmpD3Node : public CmpDNode {
   virtual uint ideal_reg() const { return Op_RegI; }
 };
 
+// Does a ScopedValue.get() hits in the cache?
+class ScopedValueGetLoadFromCacheNode;
+class ScopedValueGetHitsInCacheNode : public CmpNode {
+private:
+  // There are multiple checks involved, keep track of their profile data
+  struct {
+      float _cnt;
+      float _prob;
+  } _profile_data[3];
+
+  virtual uint size_of() const { return sizeof(*this); }
+  uint hash() const { return NO_HASH; }
+
+public:
+  enum {
+      ScopedValue = 3, // What ScopedValue object is it for?
+      Memory, // Memory for the cache loads
+      Index1, // index for the first check
+      Index2  // index for the second check
+  };
+
+  ScopedValueGetHitsInCacheNode(Compile* C, Node* c, Node* scoped_value_cache, Node* null_con, Node* mem, Node* sv,
+                                Node* index1, Node* index2) :
+          CmpNode(scoped_value_cache, null_con) {
+    init_req(0, c);
+    assert(req() == ScopedValue, "");
+    add_req(sv);
+    assert(req() == Memory, "");
+    add_req(mem);
+    assert(req() == Index1, "");
+    add_req(index1);
+    assert(req() == Index2, "");
+    add_req(index2);
+  }
+
+  Node* scoped_value() const {
+    return in(ScopedValue);
+  }
+
+  Node* mem() const {
+    return in(Memory);
```

Comment:
Why not verify that this is a `MemNode`?

Reply:
MemNodes are not the only ones carrying memory state (projection for a call, membar etc).